source.go

```diff
@@ -83,8 +82,55 @@ func (src *SourceInline) AsState(name string) (llb.State, error) {
 	return llb.Scratch().With(src.Dir.PopulateAt("/")), nil
 }
 
-func (src *SourceContext) AsState(includes []string, excludes []string, sOpt SourceOpts, opts ...llb.ConstraintsOpt) (llb.State, error) {
-	st, err := sOpt.GetContext(src.Name, localIncludeExcludeMerge(includes, excludes), withConstraints(opts))
+// withFollowPath similar to using [llb.IncludePatterns] except that it will
+// follow symlinks at the provided path.
+func withFollowPath(p string) localOptionFunc {
+	return func(li *llb.LocalInfo) {
+		if isRoot(p) {
+			return
+		}
+
+		paths := []string{p}
+		if li.FollowPaths != "" {
+			var ls []string
+			if err := json.Unmarshal([]byte(li.FollowPaths), &ls); err != nil {
+				panic(err)
+			}
+			paths = append(ls, paths...)
+		}
+		llb.FollowPaths(paths).SetLocalOption(li)
+	}
+}
+
+func (src *SourceContext) AsState(path string, includes []string, excludes []string, sOpt SourceOpts, opts ...llb.ConstraintsOpt) (llb.State, error) {
+	if !isRoot(path) {
+		// Make sure include/exclude patterns are specifically referencing paths
+		// beneath the provided path
+		if len(includes) > 0 {
+			updated := make([]string, 0, len(includes))
+			for _, p := range includes {
+				if !pathHasPrefix(p, path) {
+					p = filepath.Join(path, p)
+				}
+				updated = append(updated, p)
+			}
+			includes = updated
```

Comment:
You could probably stick this logic into a small local function and call it for both the includes and the excludes